File: src/main/java/broadcast/AccumulatorOperator.java
package main.java.broadcast;

import org.apache.spark.Accumulator;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;

/**
 * 定义在Driver端，Dirver可以读取值，操作在Executor端，在Executor是不能读取数据的。
 * @author root
 *
 */
public class AccumulatorOperator {
	public static void main(String[] args) {
		SparkConf sparkConf = new SparkConf()
				.setMaster("local")
				.setAppName("BroadCastOperator");
		final JavaSparkContext sc = new JavaSparkContext(sparkConf);
		JavaRDD<String> linesRDD = sc.textFile("cs");
		final Accumulator<Integer> countAccumulator = sc.accumulator(0);
		
		linesRDD.map(new Function<String, String>() {

			/**
			 * 
			 */
			private static final long serialVersionUID = 1L;

			@Override
			public String call(String v1) throws Exception {
				countAccumulator.add(1);
				System.out.println(countAccumulator.value());
				return v1;
			}
		}).collect();
		
		System.out.println(countAccumulator.value());
		
		
		
		sc.stop();
	}
}	

File: src/main/scala/operator/AccumulatorOperator.scala
package main.scala.operator

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext

/**
 * 注意点：
 * 	在Driver端定义，在Excutor端操作，不能再Executor端读取
 */
object AccumulatorOperator {
  def main(args: Array[String]): Unit = {
     val conf = new SparkConf().setAppName("AccumulatorOperator").setMaster("local")
    val sc = new SparkContext(conf)

     val rdd = sc.textFile("F:\\JAVA\\JavaWorkSpace\\testmaven\\src\\main\\resources\\cs")
     
      val count = sc.accumulator(0)
    
     rdd.foreach { x => {
       count.add(1)
//       println(count.value)
     } }
      
     println(count.value)
     
     sc.stop()
    
    
    
  }
}